session-replay/Tests/DatadogSessionReplayTests/Recorder/ViewTreeSnapshotProducer/Utilties/UIImageView+RecordingTests.swift

```diff
@@ -0,0 +1,173 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2019-2020 Datadog, Inc.
+ */
+
+import XCTest
+import UIKit
+@testable import DatadogSessionReplay
+
+/// Make sure to run on iPhone 11
+class UIImageViewRecordingTests: XCTestCase {
+    var frame: CGRect = .zero
+    var imageView: UIImageView = .init()
+
+    override func setUp() {
+        super.setUp()
+        frame = CGRect(x: 10, y: 10, width: 100, height: 100)
+        imageView = UIImageView(frame: frame)
+        if #available(iOS 13.0, *) {
+            imageView.image = UIImage(systemName: "testtube.2")
+        }
+    }
+
+    override func tearDown() {
+        imageView = .init()
+        frame = .zero
+        super.tearDown()
+    }
+
+    func testImageFrameForScaleAspectFit() {
+        imageView.contentMode = .scaleAspectFit
+
+        let imageFrame = imageView.imageFrame(in: frame)
+        XCTAssertEqual(imageFrame, CGRect(x: 10.0, y: 13.57142857142857, width: 100.0, height: 92.85714285714286))
```

Comment:
When running on iPhone 14 Pro simulator (iOS 16.2) I get this:
```
XCTAssertEqual failed: ("(10.0, 13.174603174603178, 100.0, 93.65079365079364)") is not equal to ("(10.0, 13.57142857142857, 100.0, 92.85714285714286)")
```
which is obviously because I didn't follow the `/// Make sure to run on iPhone 11` comment. However, as annotated in another comment, we cannot introduce such strong dependency on certain environment.

On one side, the `13.174603174603178` vs `13.57142857142857` difference could be solved if using `accuracy`:
```swift
XCTAssertEqual(imageFrame.y, 13.57, accurracy: 0.01)
```
but this won't work reliably for `93.65` vs `92.85`.

That makes me think if we should depend on the actual `UIImageView` in tests. I could imagine scaling logic being implemented as type-agnostic, pure math function similar to [AVMakeRect(aspectRatio:insideRect:)](https://developer.apple.com/documentation/avfoundation/1390116-avmakerect):
```swift
func rect(aspectRatio: CGSize, insideRect: CGRect, contentMode: UIView.ContentMode) -> CGRect
```
If such is defined independently (not as an `UIImageView` extension), we will be able to test it on deterministic values, not influenced by any simulator or runtime environment. The `UIImageView.imageFrame(in:)` could then become a convenience to use it for images, but it won't be that critical to assert its precise values in tests.

How does it sound @maciejburda?

Reply:
Interesting thoughts. I was thinking about this problem when I realised it gives different values for different simulators, although - I'm not sure why it's happening. In the test I'm creating a square UIImageView and provide explicit frame. My guess is that it's related to different pixel density of screens.

One thought I had - is that maybe we could leverage snapshot tests for things like this, but this would for sure need a lock on the simulator type and version. These unit tests don't really help much in terms of understanding what really broke, whereas snapshots would show clear visual representation of what has changed.

My initial implementation was relying on [AVMakeRect(aspectRatio:insideRect:)](https://developer.apple.com/documentation/avfoundation/1390116-avmakerect): but as far as I could tell it was producing the same results as the function written. I will explore our options here a little bit more and let you know.